core/sr-io/without_std.rs

```diff
@@ -624,10 +732,200 @@ impl CryptoApi for () {
+		let mut len = 0u32;
+		unsafe {
+			let ptr = ext_local_storage_get.get()(
+				key.as_ptr(),
+				key.len() as u32,
+				&mut len,
+			);
+
+			if len == u32::max_value() {
+				None
+			} else {
+
+				// Invariants required by Vec::from_raw_parts are not formally fulfilled.
+				// We don't allocate via String/Vec<T>, but use a custom allocator instead.
+				// See #300 for more details.
+				Some(<Vec<u8>>::from_raw_parts(ptr, len as usize, len as usize))
+			}
+		}
+	}
+
+	fn http_request_start(method: &str, url: &str, meta: &[u8]) -> Result<offchain::HttpRequestId, ()> {
+		let method = method.as_bytes();
+		let url = url.as_bytes();
+
+		unsafe {
+			let result = ext_http_request_start.get()(
+				method.as_ptr(),
+				method.len() as u32,
+				url.as_ptr(),
+				url.len() as u32,
+				meta.as_ptr(),
+				meta.len() as u32,
+			);
+
+			if result > u16::max_value() as u32 {
+				Err(())
+			} else {
+				Ok(offchain::HttpRequestId(result as u16))
+			}
+		}
+	}
+
+	fn http_request_add_header(request_id: offchain::HttpRequestId, name: &str, value: &str) -> Result<(), ()> {
+		let name = name.as_bytes();
+		let value = value.as_bytes();
+
+		unsafe {
+			let result = ext_http_request_add_header.get()(
+				request_id.0 as u32,
+				name.as_ptr(),
+				name.len() as u32,
+				value.as_ptr(),
+				value.len() as u32,
+			);
+
+			if result == 0 {
+				Ok(())
+			} else {
+				Err(())
+			}
+		}
+	}
+
+	fn http_request_write_body(
+		request_id: offchain::HttpRequestId,
+		chunk: &[u8],
+		deadline: Option<offchain::Timestamp>
+	) -> Result<(), ()> {
+		let res = unsafe {
+			ext_http_request_write_body.get()(
+				request_id.0 as u32,
+				chunk.as_ptr(),
+				chunk.len() as u32,
+				deadline.map_or(0, |x| x.unix_millis()),
+			)
+		};
+
+		if res == 0 { Ok(()) } else { Err(()) }
+	}
+
+	fn http_response_wait(
+		ids: &[offchain::HttpRequestId],
+		deadline: Option<offchain::Timestamp>
+	) -> Vec<offchain::HttpRequestStatus> {
+		let ids = ids.iter().map(|x| x.0 as u32).collect::<Vec<_>>();
+		let mut statuses = Vec::new();
+		statuses.resize(ids.len(), 0u32);
+
+		unsafe {
+			ext_http_response_wait.get()(
+				ids.as_ptr(),
+				ids.len() as u32,
+				statuses.as_mut_ptr(),
+				deadline.map_or(0, |x| x.unix_millis()),
+			)
+		}
+
+		statuses
+			.into_iter()
+			.map(|status| offchain::HttpRequestStatus::from_u32(status).unwrap_or_else(|| offchain::HttpRequestStatus::Unknown))
+			.collect()
+	}
+
+	fn http_response_headers(
+		request_id: offchain::HttpRequestId,
+	) -> Vec<(Vec<u8>, Vec<u8>)> {
+		let mut len = 0u32;
+		let raw_result = unsafe {
+			let ptr = ext_http_response_headers.get()(
+				request_id.0 as u32,
+				&mut len,
+			);
+			// Invariants required by Vec::from_raw_parts are not formally fulfilled.
+			// We don't allocate via String/Vec<T>, but use a custom allocator instead.
+			// See #300 for more details.
+			<Vec<u8>>::from_raw_parts(ptr, len as usize, len as usize)
```

Comment:
This is really subtle. `ext_http_response_headers` potentially can invoke UB here, if `ptr` is `null` here, even if `len` equals 0.

We might want to either:

- explicitly mention this here and for the documentation in `ext_http_response_header` and guarantee it
- guard here against this explicitly here.